read_buffer/src/column/encoding/scalar/rle.rs

```diff
@@ -0,0 +1,1077 @@
+/// stored as a tuple: `(run_length, value)`, where `run_length` indicates the
+/// number of times the value is to be repeated.
+#[derive(Debug, Default)]
+pub struct RLE<T>
+where
+    T: PartialOrd + Debug,
+{
+    // stores tuples of run-lengths. Each value is repeats the total number of
+    // times the second value should is repeated within the column.
+    //
+    // TODO(edd): can likely improve on 4B for run-length storage
+    //
+    // TODO(edd): Option<T> effectively doubles the size of storing `T` for many
+    //            `T`. I am not overly concerned about this right now because of
+    //            the massive savings the encoding can already provide. Further
+    //            when `T` is a `NonZeroX` then we get the niche optimisation
+    //            that makes `Option<T>` the same size as `T`.
+    //
+    //            For floats I plan to implement a check that will allow us to
+    //            store them as integers if they're natural numbers, then we can
+    //            make use of `Option<NonZeroX>` with no option overhead. If
+    //            they're not all natural numbers then I think we could make a
+    //            `NonZeroNaN` and perhaps figure out how to get `Option<NonZeroNan>`
+    //            to have no overhead.
+    //
+    // TODO(edd): another obvious thing that might be worth doing in the future
+    //            is to store all the run-lengths contiguously rather than in
+    //            `Vec<(,)>`. One advantage of this is that it is possible to
+    //            avoid a tuple "overhead" when there is only a single
+    //            occurrence of a value by using a specific marker to represent
+    //            when something is a run-length or something is just a single
+    //            value.
+    run_lengths: Vec<(u32, Option<T>)>,
+
+    // number of NULL values in the column
+    null_count: u32,
+
+    // number of total logical rows (values) in the columns.
+    num_rows: u32,
+}
+
+impl<T> std::fmt::Display for RLE<T>
+where
+    T: std::fmt::Debug + Display + PartialOrd + Copy,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "[{}] size: {:?} rows: {:?} nulls: {} runs: {} ",
+            self.name(),
+            self.size(),
+            self.num_rows(),
+            self.null_count(),
+            self.run_lengths.len()
+        )
+    }
+}
+
+impl<T: PartialOrd + Debug + Copy> RLE<T> {
+    /// The name of this encoding.
+    pub fn name(&self) -> &'static str {
+        ENCODING_NAME
+    }
+
+    /// A reasonable estimation of the on-heap size this encoding takes up.
+    pub fn size(&self) -> usize {
+        size_of::<Vec<(u32, Option<T>)>>() // run length container size
+            + (self.run_lengths.len() * size_of::<(u32, Option<T>)>()) // run lengths
+            + size_of::<u32>()+size_of::<u32>() // null count, num rows
```

Comment:
We could simplify this to:

```suggestion
        std::mem::size_of::<Self>() + (self.run_lengths.len() * size_of::<(u32, Option<T>)>())
```

Which is also a little easier to maintain - `size_of::<Self>()` computes the overhead of all the fields in the struct (`Vec` container & counters) and still takes into account the concrete [sizes of T](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=001b12dc6327a0ca39a3e5427232bbb5):